src/icon-api.js

```diff
@@ -0,0 +1,96 @@
+export default class IconApi {
+  #dataType
+  #dataStore
+
+  /**
+   * @param {Object} opts
+   * @param {import('./datatype/index.js').DataType<
+   *   import('./datastore/index.js').DataStore<'config'>,
+   *   typeof import('./schema/project.js').iconTable,
+   *   'icon',
+   *   import('@mapeo/schema').Icon,
+   *   import('@mapeo/schema').IconValue
+   * >} opts.iconDataType
+   * @param {import('./datastore/index.js').DataStore<'config'>} opts.iconDataStore
+   */
+  constructor({ iconDataType, iconDataStore }) {
+    this.#dataType = iconDataType
+    this.#dataStore = iconDataStore
+  }
+
+  /**
+   * @param {Omit<import('@mapeo/schema').IconValue, 'variants'> &
+   * {variants: Array<
+   *    import('@mapeo/schema/dist/proto/icon/v1.js').Icon_1_IconVariant
+   *    & {blob: Buffer}>}} icon
+   */
+  async create(icon) {
+    if (icon.variants.length < 1) {
+      throw new Error('empty variants array')
+    }
+    const blobsIds = await Promise.all(
+      icon.variants.map(({ blob }) => {
+        return this.#dataStore.writeRaw(blob)
+      })
+    )
+    const variantsWithId = icon.variants.map((variant, i) => {
+      // eslint-disable-next-line no-unused-vars
+      const { blob, ...variantWithoutBlob } = variant
+      return { ...variantWithoutBlob, blobVersionId: blobsIds[i] }
+    })
+    return await this.#dataType.create({
+      schemaName: 'icon',
+      name: icon.name,
+      // @ts-ignore TODO: remove
+      variants: variantsWithId,
+    })
+  }
+
+  /** @typedef {'image/png' | 'image/svg+xml'} ValidMimeType */
+
+  /**
+   * @param {Object} opts
+   * @param {String} opts.iconId
+   * @param {String} [opts.size]
+   * @param {number} [opts.pixelDensity]
+   * @param {ValidMimeType} [opts.mimeType]
+   */
+  async getIcon({ iconId, size, pixelDensity, mimeType }) {
+    const iconRecord = await this.#dataType.getByDocId(iconId)
+    const iconVariant = this.#getBestVariant(iconRecord.variants, {
+      size,
+      pixelDensity,
+      mimeType,
+    })
+    return await this.#dataStore.readRaw(iconVariant.blobVersionId)
+  }
+
+  /**
+   * @param {import('@mapeo/schema').IconValue['variants']} variants
+   * @param {object} opts
+   * @param {string} [opts.size]
+   * @param {number} [opts.pixelDensity]
+   * @param {ValidMimeType} [opts.mimeType]
+   **/
+  #getBestVariant(variants, { size, pixelDensity, mimeType }) {
```

Comment:
What you say makes sense, the implementation doesn't exactly do that from what I can see:

- If the mimeType does not match, in this implementation you will still get back the first variant (we should probably throw / return not found)
- We probably want the variant that is closest in terms of size and pixel density. With the current implementation anything that has one thing matching (e.g. match score 1) will replace the previous one as the "choice", and not sure that is always the best choice?

This is probably worth unit testing - it's the kind of function that could easily have bugs that are hard to track down in other tests.

Reply:
Ok. Just to clarify the criteria for this - to see if I'm getting it right:
* Matching mimeType is required, so if asking for something with `mimeType: image/png` that doesn't have a match, we should throw an Error
* regarding `size` and `pixelDensity`, are the two equally important? (in terms of code, do they add the same score when looking for the best match?)
* What happens with a tie? (i.e. we ask for `{pixelDensity:1, size:2}` and the variants that exist are  `[{pixelDensity: 1, size:1}, {pixelDensity: 2, size: 1}]`, which means each variant have one match, resulting en equal score?). We can flip a coin (`Math.random() > 0.5`) ....

Reply:
Mh, I realize that flipping a coin shouldn't be a possibility - more so with testing... -. We should probably choose between:
1. Weighting the score of `pixelDensity` and `size` differently (would this eliminate ties??)
2. Have a predictable behaviour in case of a tie (choose first? choose last?)